lib_eio/sem_state.ml

```diff
@@ -0,0 +1,189 @@
+   handle it.
+
+   If the cancelling user sees the Resumed state after Cancelling then it knows
+   that the resuming user has transferred to it the responsibility of waking
+   one user. If the cancelling user is also responsible for waking one user
+   then it performs an extra resume on behalf of the resuming user.
+
+   Finally, if the cancelling user is not responsible for waking anyone (even
+   itself) then it leaves the cell in Cancelling (the CQS paper uses a separate
+   Refused state, but we don't actually need that). This can only happen when a
+   resume is happening at the same time. The resumer will transition to
+   Resumed, creating an obligation to resume, but we've just done that anyway.
+   We know this Cancelling state can't last long because at the moment when the
+   canceller incremented the counter all current waiters, including itself,
+   were in the process of being resumed. *)
+
+module Cell = struct
+  type _ t =
+    | Request of (unit -> unit)         (* Waiting for a resource *)
+    | Cancelling                        (* Cancellation is in progress (or refused) *)
+    | Cancelled                         (* Ignore this and use next cell instead *)
+    | Resumed                           (* Suspender now owns resource *)
+    | Empty
+
+  let init = Empty
+
+  let segment_order = 2
+
+  let dump f = function
+    | Request _ -> Fmt.string f "Request"
+    | Empty -> Fmt.string f "Empty"
+    | Resumed -> Fmt.string f "Resumed"
+    | Cancelling -> Fmt.string f "Cancelling"
+    | Cancelled -> Fmt.string f "Cancelled"
+end
+
+module Cells = Cells.Make(Cell)
+
+type cell = unit Cell.t
+
+type t = {
+  state : int Atomic.t;         (* Free resources. Negative if there are waiters waiting. *)
+  cells : unit Cells.t;
+}
+
+type request = t * unit Cells.segment * unit Cell.t Atomic.t
+
+(* Wake one waiter. *)
+let rec resume t =
+  let cell = Cells.next_resume t.cells in
+  let rec aux () =
+    match (Atomic.get cell : cell) with
+    | Request r as cur ->
+      (* The common case: we have a waiter for the value *)
+      if Atomic.compare_and_set cell cur Resumed then r ()
+      else aux ()
+    | Empty ->
+      (* The consumer has reserved this cell but not yet stored the request.
+         We place Resumed there and the consumer will handle it soon. *)
+      if not (Atomic.compare_and_set cell Empty Resumed) then aux ()
+    | Cancelled ->
+      (* This waker has finished cancelling. Ignore it and resume the next one. *)
+      resume t
+    | Cancelling ->
+      (* The waker has started cancelling. Let it know we want to resume it
+         and then let it handle it (it may also have finished cancelling but
+         decided there's nothing left to do anyway). *)
+      if not (Atomic.compare_and_set cell Cancelling Resumed) then aux ()
+    | Resumed ->
+      (* This state is unreachable because we (the provider) haven't set this yet *)
+      assert false
+  in
+  aux ()
```

Comment:
I do not yet fully understand the `Cells` abstraction.  However, looking at the code here it seems that `Resumed` is a terminal state.  If that is the case, then it may be safe to always exchange the state to `Resumed`.  Like I said, I do not yet fully understand `Cells` so this might be wrong.

However, assuming that exchanging to `Resumed` is a safe, then the above could be changed to eliminate the loop:

```ocaml
  let rec resume t =
    let cell = Cells.next_resume t.cells in
    match (Atomic.exchange cell Resumed : cell) with
    | Request r ->
      (* The common case: there was a waiter for the value *)
      r ()
    | Cancelled ->
      (* The waker had finished cancelling. Ignore it and resume the next one. *)
      resume t
    | (Empty
       (* The consumer had reserved this cell but not yet stored the request.
          We placed Resumed there and the consumer will handle it soon. *)
      | Cancelling
        (* The waker had started cancelling. We let it know we want to resume it
           and then let it handle it. *)
      | Resumed
        (* We lost the race, just ignore *)) ->
      ()
```

Assuming it is safe to always set to `Resumed`, the above version avoids the loop overhead and only does a single atomic read-write.  Assuming there are not many racing to do the same this may be slightly faster. This also likely results in a smaller amount of machine code.

Reply:
> This also likely results in a smaller amount of machine code.

We could combine the `Empty` and `Cancelled` handlers to save some code though (they're identical). But maybe it's clearer as it is.

Reply:
Let me just play a devil's advocate a bit here. :)

Perhaps one possibility might be to merge `Resumed` and `Cancelled` into a single, let's say, `Terminal` state.  There is a race to set the cell to `Terminal` state.  The action depends on the state from which the `Terminal` state was reached.

> One side-effect is that calling cancel twice currently reliably raises Invalid_argument

About `cancel`.  Hmm... Under which circumstances one might call `cancel` twice?  (Asking because I'm not sure I understand the concern.)

Assuming there is a race to call `cancel`, I would tend to think that it is important that `cancel` returns `true` at most once.  This way any follow up actions to successful cancellation would only be performed once.  If this makes sense, then the API might be more user friendly without raising Invalid_argument.  John Ousterhout talks about this in his talk on software design [here](https://youtu.be/bmSAYlu0NcY?t=1312).